File: Two Pointers/LC_167_Two_Sum_II.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.17

class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        l, r = 0, len(numbers) - 1
        while l < r:
            temp = numbers[l] + numbers[r]
            if temp == target:
                return [l + 1, r+1]
            if temp > target:
                r -= 1
            else:
                l += 1
        """
        Hashmap version
        mapping = collections.defaultdict(int)
        for idx, num in enumerate(numbers):
            if target - num in mapping:
                return sorted([idx+1, mapping[target-num] + 1])
            mapping[num] = idx
        
        """

File: Two Pointers/LC_167_Two_Sum_II.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.17

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        // use two pointers, this can be extended to kSum!
        vector<int> ans;
        int i = 0, j = numbers.size() - 1;
        while (true){
            int cur = numbers[i] + numbers[j];
            if (cur == target){
                ans.push_back(i+1);
                ans.push_back(j+1);
                break;
            }
            else if (cur > target){
                j--;
            }
            else{
                i++;
            }
        }
        return ans;
    }
};